src/Microsoft.Health.Fhir.Core/Features/Validation/NarrativeValidator.cs

```diff
@@ -31,50 +30,51 @@ public override IEnumerable<ValidationFailure> Validate(PropertyValidatorContext
 
             if (context.PropertyValue is ResourceElement resourceElement)
             {
-                var poco = resourceElement.ToPoco();
-
-                if (poco is DomainResource resource)
+                if (resourceElement.IsDomainResource)
                 {
-                    foreach (ValidationFailure validationFailure in ValidateResource(resource))
+                    foreach (ValidationFailure validationFailure in ValidateResource(resourceElement))
                     {
                         yield return validationFailure;
                     }
                 }
-
-                if (poco is Bundle bundle)
+                else if (resourceElement.InstanceType.Equals(KnownResourceTypes.Bundle, System.StringComparison.OrdinalIgnoreCase))
                 {
-                    var domainResources = bundle.Entry.Select(x => x.Resource).OfType<DomainResource>();
-
-                    foreach (ValidationFailure validationFailure in domainResources.SelectMany(ValidateResource))
+                    var bundleEntries = resourceElement.Instance.Select(KnownFhirPaths.BundleEntries);
+                    if (bundleEntries != null)
                     {
-                        yield return validationFailure;
+                        var domainResources = bundleEntries.Select(e => new ResourceElement(e)).Where(r => r.IsDomainResource);
```

Comment:
In thinking about it, perhaps we don't care about the `IsDomainResource` at all, and just simply filter on something like `DomainResource.text.div.exists()` in the where clause? If the resource does not contain text, then we can skip it anyway

The path `DomainResource.text.div` could remain in the KnownPaths class

Reply:
Scalar lookup using the DomainResource.text.div throws an exception if we try it on a resource which is not a DomainResource.  I think I will leave the Where clause in as it makes it very explicit that we only want to check this on DomainResources.